### DISCUSSION TITLE
Fail to create tunnel interface on Android with Multiple User

### DISCUSSION THREAD

**Discussion by etlweather** [strongswan/strongswan#1930] [General]
Hello, I am wondering if anyone encountered (and solved) this issue and/or if there is anything I can do... I have an Android 10 mobile barcode scanner HT730 from Unitech. It is set up with multiple-users. Trying to bring up the VPN with StrongSwan 2.4.2, I get this error:

```
10-01 20:30:40.763   663  1270 D NetlinkEvent: Unknown ifindex 19 in RTM_DELADDR
10-01 20:30:40.835  2949  5732 W System.err: java.lang.SecurityException: getPackageUid: Neither user 10170 nor current process has android.permission.INTERACT_ACROSS_USERS.
10-01 20:30:40.835  2949  5732 W System.err:    at android.os.Parcel.createException(Parcel.java:2071)
10-01 20:30:40.835  2949  5732 W System.err:    at android.os.Parcel.readException(Parcel.java:2039)
10-01 20:30:40.835  2949  5732 W System.err:    at android.os.Parcel.readException(Parcel.java:1987)
10-01 20:30:40.835  2949  5732 W System.err:    at android.net.IConnectivityManager$Stub$Proxy.establishVpn(IConnectivityManager.java:3028)
10-01 20:30:40.835  2949  5732 W System.err:    at android.net.VpnService$Builder.establish(VpnService.java:897)
10-01 20:30:40.835  2949  5732 W System.err:    at org.strongswan.android.logic.CharonVpnService$BuilderAdapter.establishIntern(CharonVpnService.java:921)
10-01 20:30:40.835  2949  5732 W System.err:    at org.strongswan.android.logic.CharonVpnService$BuilderAdapter.establish(CharonVpnService.java:946)
10-01 20:30:40.835  2949  5732 I charon  : 13[LIB] builder: failed to build TUN device
10-01 20:30:40.835  2949  5732 I charon  : 13[DMN] failed to setup TUN device
```

Anybody has an idea of a solution? I searched quite a bit and couldn't find anything yet.

**Comment by etlweather** [comment]
I did try turning off multi-user but unfortunately get the same error.

**Comment by tobiasbrunner** [comment]
The app calls `VpnService.Builder::establish()`, which is an Android API call. The error seems to be caused by that, so the app has no direct control over it.

From what I've seen regarding this error, it seems to be caused if apps are excluded from the VPN (`addDisallowedApplication()`) on systems with multiple users where one of them is restricted. The problem is that the app excludes itself from the VPN to allow it to fetch CRLs and OCSP through the blocking TUN device we install. To me it seems to be a system bug that the VpnService API is affected by this.

It might be possible to explicitly grant the permission (which is a signature/system-level one a regular app can't request itself) via `adb`:

    adb shell pm grant org.strongswan.android android.permission.INTERACT_ACROSS_USERS

But it might have to be requested in the app's manifest before this works.

Deleting the restricted account might also work (although, you mentioned that you disabled "multi-user", so not sure if that's included in that).

**Reply by etlweather** [reply]
> It might be possible to explicitly grant the permission (which is a signature/system-level one a regular app can't request itself) via adb: 
>    
> But it might have to be requested in the app's manifest before this works.

Here is the output from this command:

```
# adb shell pm grant org.strongswan.android android.permission.INTERACT_ACROSS_USERS

Security exception: Package org.strongswan.android has not requested permission android.permission.INTERACT_ACROSS_USERS

java.lang.SecurityException: Package org.strongswan.android has not requested permission android.permission.INTERACT_ACROSS_USERS
```

P.S.: Thanks for checking into this!

**Reply by etlweather** [reply]
> Deleting the restricted account might also work (although, you mentioned that you disabled "multi-user", so not sure if that's included in that).

Ah! Thanks for the prompt! 

I re-enabled multiple users and deleted the additional user that had been created (that leaves 1 user - owner). I then disabled multiple users and the VPN connected!

**Reply by tobiasbrunner** [reply]
> Here is the output from this command:
> 
> ```
> # adb shell pm grant org.strongswan.android android.permission.INTERACT_ACROSS_USERS
> 
> Security exception: Package org.strongswan.android has not requested permission android.permission.INTERACT_ACROSS_USERS
> 
> java.lang.SecurityException: Package org.strongswan.android has not requested permission android.permission.INTERACT_ACROSS_USERS
> ```

Thanks for testing. So this might actually work if we added that to the manifest. Maybe we could do that in a future update of the app.

> I re-enabled multiple users and deleted the additional user that had been created (that leaves 1 user - owner). I then disabled multiple users and the VPN connected!

OK, great. At least we know a workaround (obviously not ideal if a user relies on the multi-user feature).
